File: Data Structures/Tree-based-DSAs/Binary-tree/binary-tree.cpp
//binary tree cpp implementation:

#include "stdlib.h"
#include "iostream"

using namespace std;

struct Node {
    int data;
    struct Node * left = nullptr;
    struct Node * right = nullptr;
};

struct Node * newNode(int data){
    struct Node *node = (struct Node *)malloc(sizeof(struct Node));
    node->data = data;
    node->left = nullptr;
    node->right = nullptr;

    return node;
}

//traverse preorder:

void traversePreorder(struct Node * root){
    if (root){
        cout << root->data << " ";
        traversePreorder(root->left);
        traversePreorder(root->right);
    }
}

//traverse inorder:
void traverseInorder(struct Node * root){
    if (root){
        traverseInorder(root->left);
        cout << root->data << " ";
        traverseInorder(root->right);
    }
}

//traverse postorder:
void traversePostorder(struct Node * root){
    if (root){
        traversePostorder(root->left);
        traversePostorder(root->right);
        cout << root->data << " ";
    }
}

//checking whether the binary tree is a full binary tree:

bool isFullBinaryTree(struct Node *root){

    //is binary tree empty?
    if (!root){
        return true;
    }

    //check whether a child is present:
    if ((!root->left) && (!root->right)){
        return true;
    }

    //checking internal nodes:
    if ((root->left) && (root->right)){
        return (isFullBinaryTree(root->left) && isFullBinaryTree(root->right));
    }

    //else return false:
    return false;
}

//calculating the height of the tree:
int height(Node * root){
    int d = 0;
    while (root){
        d +=1;
        root = root->left;
    }
    return d;
}

//checking whether the binary tree is a perfect binary tree:
bool isPerfectR(Node *root, int d, int level = 0){
    //is the binary tree empty?
    if (!root){
        return true;
    }

    if (!root->left && !root->right){
        return (d == level + 1);
    }

    if (!root->left || !root->right){
        return false;
    }

    return (isPerfectR(root->left, d, level + 1 ) && (isPerfectR(root->right, d, level + 1)));
    
}
//count the number of nodes in the tree:

int countNodes(Node * root){
    if (!root){
        return 0;
    }
    return (1 + countNodes(root->left) + (countNodes(root->right)));
}

//check for a complete binary tree:
bool isCompleteBinaryTree(Node * root, int index, int nodeCount){
    if (!root){
        return true;
    }

    if (index >= nodeCount){
        return false;
    }

    return (isCompleteBinaryTree(root->left, 2 * index + 1, nodeCount) && 
            isCompleteBinaryTree(root->right, 2 * index + 2, nodeCount));
}

bool isPerfect(Node *root){
    int d = height(root);
    return isPerfectR(root, d);
}


int main(){
    struct Node * root = newNode(20);
    root->left = newNode(30);
    root->right = newNode(40);
    root->left->left = newNode(70);

    cout <<"Traverse inorder ->" << " ";
    traverseInorder(root);

    cout <<"\nTraverse preorder ->" << " ";
    traversePreorder(root);

    cout << "\nTraverse postorder ->" << " ";
    traversePostorder(root);

    cout << "\n is this a full binary tree? " << (isFullBinaryTree(root) ? "yes": "no") << " ";
    cout <<"\n is this a perfect binary tree? " << (isPerfect(root) ? "yes": "no") << " ";
    cout << "\n is this a complete binary tree? " << (isCompleteBinaryTree(root, 0, countNodes(root)) ? "yes": "no");



}

File: Data Structures/Tree-based-DSAs/Binary-tree/binary-tree.py
#binary tree application in python:

""" This is the node structure and traversal methods: """
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def traversePreorder(self):
        print(str(self.data) + "->", end=" ")

        if self.left:
            self.left.traversePreorder()
        if self.right:
            self.right.traversePreorder()
    def traversePostorder(self):
        if self.left:
            self.left.traversePostorder()
        if self.right:
            self.right.traversePostorder()
        print(str(self.data) + "->", end=" ")

    def traverseInorder(self):
        if self.left:
            self.left.traverseInorder()
        print(str(self.data) + "->", end=" ")
        if self.right:
            self.right.traverseInorder()

""" check for full binary tree: """

def isFullBtree(root):

    #1. if tree is empty:
    if root is None:
        return True
    #2. checking whether a child is present:
    if root.left is None and root.right is None:
        return True
    if root.left is not None and root.right is not None:
        return (isFullBtree(root.left) and isFullBtree(root.right))
    return False

def calculateheight(node):
    d = 0
    while (node is not None):
        d += 1
        node = node.left
    return d

def isPerfectBinaryTree(root, d, level = 0):
    #is the tree empty?
    if (root is None):
        return True
    
    #check the presence of trees:
    if (root.left is None and root.right is None):
        return (d == level + 1)

    if (root.left is None or root.right is None):
        return False
    return (isPerfectBinaryTree(root.left, d , level + 1)) and (isPerfectBinaryTree(root.right, d, level + 1))

def isCompleteBinaryTree(root, index, numberOfNodes):
    #is the tree empty?

    if root is None:
        return True
    if index >= numberOfNodes:
        return False
    return (isCompleteBinaryTree(root.left, 2 * index + 1, numberOfNodes) 
            and (isCompleteBinaryTree(root.right, 2 * index + 2, numberOfNodes)))

#do this as python does not have references:

class Height:
    def __init__(self):
        self.height = 0
        
def isHeightBalanced(root, height):
    left_height = Height()
    right_height = Height()

    if root is None:
        return True
    l = isHeightBalanced(root.left, left_height)
    r = isHeightBalanced(root.right, right_height)

    height.height = max(left_height.height, right_height.height) + 1

    if abs(left_height.height - right_height.height) <= 1:
        return l and r
    return False

    

    
def count_nodes(root):
    if root is None:
        return 0
    return (1 + count_nodes(root.left) + (count_nodes(root.right))) 


if __name__ == "__main__":
    binTree = Node(20)
    binTree.left = Node(30)
    binTree.right = Node(40)
    binTree.left.left = Node(50)
    # binTree.left.left.left = Node(70)

    print("Traversing inorder->")
    binTree.traverseInorder()
    print()

    print("Traversing preorder->")
    binTree.traversePreorder()
    print()

    print("Traversing postorder->")
    binTree.traversePostorder()
    print()

    print("Is this a full tree? -> " + ("yes" if isFullBtree(binTree) else "no"))
    print("Height of the tree:-> ", calculateheight(binTree))
    print("Is this a perfect binary tree? -> " + ("yes" if (isPerfectBinaryTree(binTree, calculateheight(binTree))) else "no"))

    nodeCount = count_nodes(binTree)
    index = 0

    print("This is the node count -> ", nodeCount)
    print()
    print("Is this a complete binary tree?-> " + ("yes" if (isCompleteBinaryTree(binTree, index, nodeCount)) else "no"))
    print()
    height = Height()
    print("Is this a balanced tree? -> " + ("yes" if (isHeightBalanced(binTree, height)) else "No"))







